src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -60,6 +60,18 @@ StoragePtr InterpreterInsertQuery::getTable(ASTInsertQuery & query)
     {
         const auto & factory = TableFunctionFactory::instance();
         TableFunctionPtr table_function_ptr = factory.get(query.table_function, getContext());
+
+        /// If table function needs structure hint from select query
+        /// we can create a temporary pipeline and get the header.
+        if (query.select && table_function_ptr->needStructureHint())
+        {
+            InterpreterSelectWithUnionQuery interpreter_select{
```

Comment:
Maybe it will lead to extra execution of scalar subqueries?
Nevertheless, the benefits of this feature outpays everything :)